README

DBD::JDBC - JDBC proxy driver for the DBI module

  Copyright 1999-2001,2005-2006 Vizdom Software, Inc. All Rights Reserved.
  
  This program is free software; you can redistribute it and/or 
  modify it under the same terms as the Perl Kit, namely, under 
  the terms of either:
  
      a) the GNU General Public License as published by the Free
      Software Foundation; either version 1 of the License, or 
      (at your option) any later version, or
  
      b) the "Artistic License" that comes with the Perl Kit.
  
  This program is distributed in the hope that it will be seful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
  either the GNU General Public License or the Artistic License
  for more details. 


SUMMARY

  This module allows you to use DBI to access a JDBC driver from
  Perl and is implemented in a fashion similar to that of
  DBD::Proxy. There are two components: a pure Perl DBD::JDBC
  module, and a pure Java server application. Both must be
  installed in order to use this driver, though, like DBD::Proxy,
  they may be installed on separate machines. 

  The DBD::JDBC Java server process is threaded and supports
  multiple simultaneous connections.

  The DBD::JDBC Java server uses the following JDBC 2.0 features:
  java.sql.Types.{BLOB, CLOB, ARRAY},
  ResultSet.getCharacterStream, ResultSet.getBlob and the
  java.sql.Blob type, ResultSet.getClob and the java.sql.Clob
  type. 

  Public methods of the JDBC Connection, Statement, ResultSet,
  and ResultSetMetaData classes, including any public
  driver-specific methods not part of the JDBC API, are exposed
  using reflection and the $h->jdbc_func method. Only methods
  with primitive or String parameters and return types are
  currently supported in this way.

  This module also allows a Java application or servlet to create
  a JDBC connection and then exec a Perl script which can use
  that pre-existing JDBC connection via DBI. 

  See the JDBC.pod file for more information on usage.


PREREQUISITES

  The following must be installed before attempting to install
  this driver. 

      Perl 5.8 or higher
      Convert::BER 1.31
      DBI 1.48 or higher

  
  In addition to Perl, the following must be available in order
  to run this driver. 

      Java Virtual Machine compatible with JDK 1.4
      JDBC driver for your database of choice
      log4j-1.2.13 (included)


INSTALLATION

  There are three components to install: Convert::BER, DBD::JDBC,
  and the Java server.

  * Convert::BER 1.31
    
    See the README file included with Convert::BER for
    installation instructions.


  * The DBD::JDBC server component

    The Java classes and their source code are provided in
    dbd_jdbc.jar. Copy this file to a location of your
    choice. See the DBD::JDBC documentation (JDBC.pod) for
    instructions on running the server.

    You can install the DBD::JDBC Perl module and test it before
    setting up this server.


  * DBD::JDBC

    The installation procedure for the Perl module follows the
    usual Perl installation procedure, but see below for details
    on running the "make test" step successfully.

      perl Makefile.PL
      make
      make test
      make install

    The "make test" step will run tests against the included
    hsqldb database. The classpath and JDBC URL are preconfigured
    to allow these tests to succeed. 

    If the java executable is not in your path, set the following
    environment variable:  

      DBDJDBC_JAVA_BIN: the Java executable. This will default to
      "java", so if java is in your path, you need not set
      this. Otherwise, set it to the complete path to your java
      executable. This will be used to start the Java server
      portion of this driver.

    You may override the port which the Perl and Java components
    of this driver will use during testing. To do this, set the
    following environment variable: 

      DBDJDBC_PORT: The default value is 43321.

    There are some database-specific tests included; the test
    files are named for the database used. See the tests in
    t/. These tests make use of some standard default values. If
    you wish to run one of those database tests, set the
    following environment variables:

      CLASSPATH: this must include your JDBC driver and
        any supporting files required by that driver.

      DBDJDBC_URL: a valid URL for your JDBC driver. 
        
    The use of environment variables by the test scripts does not
    reflect normal usage of this driver; in general usage only
    the standard Java CLASSPATH environment variable needs to be
    set for the Java server portion of this driver. Also note
    that the test scripts attempt to start the Java server
    portion of this driver automatically; in normal usage, you
    will start the Java server by hand or in a startup script and
    leave it running (like a web server) for your scripts to
    connect to.


KNOWN BUGS

  Oracle LOBs can't be read with this driver. This requires
  either Oracle-specific functionality, a further extension to
  the $h->jdbc_func methods to support arbitrary objects, or support
  for JDBC 2.0.

  A problem has been reported reading many LONG values from Oracle. 

  
NOT YET IMPLEMENTED
  DBI-defined methods, including 
  
      DBI->data_sources('JDBC'), $dbh->data_sources

      the metadata methods $dbh->table_info, $dbh->tables,
      $dbh->type_info_all, $dbh->type_info, $dbh->column_info,
      $dbh->primary_key_info, $dbh->primary_key,
      $dbh->foreign_key_info 

      $sth->cancel

      $sth->bind_param_inout

      $dbh->clone

      $dbh->last_insert_id

      $dbh->get_info (partial as of 0.69)

  Access to JDBC-specific functionality. Some amount of JDBC
  functionality is currently exposed using $h->jdbc_func and Java
  reflection.


FOR MORE INFORMATION

  perldoc DBI

  For general DBI information and questions, see the DBI home
  page at
      http://dbi.perl.org/

  To contact the author of this module, send email to
      Gennis Emerson <gemerson@vizdom.com>


Changes
Changes for DBD::JDBC.

June 2006
        Version 0.70

        Changed the default behavior when reading long fields to
        read the entire field. Added a new property,
        jdbc_longreadall, which can be set to false to disable
        this behavior and return to the DBI standard behavior
        controlled by LongReadLen and LongTruncOk.

        Statements are now marked inactive when their DESTROY
        method is called. Thanks to Jesper Krogh for the
        suggestion. 

        Server changes: The server now uses log4j for its
        logging. 
        
        The server will attempt to close database connections
        when a connection thread is ending. Thanks to Jesper
        Krogh for the patch.
        

January 2006
        Version 0.69
 
        Added the first bit of $dbh->get_info support thanks to
        Jesper Krogh.

        Server changes: Updated the method which reads data in
        the server (handleRequest(FetchRequest aRequest)) to use
        different JDBC methods for BLOB, CLOB, and LONGVARCHAR
        data types. The server now retrieves BLOB and CLOB
        objects using getBlob and getClob. The use of the
        deprecated method getUnicodeStream to read LONGVARCHAR
        has been replaced by use of getCharacterStream.


November 2005
        Version 0.68

        Fixed a bug in the handle DESTROY methods which caused
        DESTROY to override the current value of $@. Code like
        the following: 
          eval { my $sth = $dbh->prepare(...); $sth->execute() }; 
          die $@ if $@; 
        might fail to die even when $sth->execute failed because
        $sth->DESTROY was called when the eval block was exited. 

        Added distribution tests which use hsqlbd thanks to Frank
        Tolstrup. hsqldb is a pure Java SQL database. This should
        make it possible to run "make test" without having
        another database set up.


November 2005
        Version 0.67

        This is DBD::JDBC 0.66 with a few further updates based
        on reviewing the DBI 1.48 documentation. 

        Fixed the local build process to synchronize the
        DBD::JDBC version with the server version. 

        Added support for $sth->{ParamValues}. (DBI 1.28)

        Removed locally-defined err, errstr, and state
        variables. (DBI 1.33)

        Updated NUM_OF_FIELDS behavior. (DBI 1.42)


January 2005
        Version 0.66
 
        Unofficial release, not sent to CPAN.
        Requires edits to DBI.pm.

        Updated DBD::JDBC to work with DBI 1.46. No work has been
        done to add support for DBI features added since the last
        DBI release with which this driver was certified. "Works
        with DBI 1.46" means that the test suite now passes
        again.

        This release uses DBI's install_method method to add
        jdbc_func. Use of install_method requires that the
        driver-specific prefix be registered with DBI. This has
        not been done as of DBI 1.46, so until a DBI release is
        made with the prefix registered, it is necessary to edit
        DBI.pm to add the jdbc_ prefix to $dbd_prefix_registry. 

        Scripts which used $h->func to call JDBC methods must now
        use $h->jdbc_func. Error handling with jdbc_func follows
        the standard DBI model, so it's no longer necessary to
        check for errors explicitly as it was when using func.

        Removed references to the DBI constant SQL_BIGINT, which
        was removed in an earlier DBI release. Scripts which need
        to use SQL_BIGINT should import the constant from DBD::JDBC: 

          use DBI qw(:sql_types);
          use DBD::JDBC qw(:sql_types);

        

May 2001
        Version 0.64

        Added support for arbitrary (string) JDBC connection
        properties. 

        Added support for "jdbc_" prefixes on method names used
        in $h->func() calls as required by DBI. Unprefixed method
        names are still supported in this release, though support
        may be removed in a future release.

        Changed the format of the connect message between the
        client and server. This means that this version of the
        client is not compatible with any older version of the
        server. 


June 2000
        Version 0.63

        Changed required version of DBI to 1.13 in order to have
        access to the SQL_BIT constant (used to specify boolean
        method parameters for func).

        Added $h->func implementations for $dbh and $sth
        objects. This implementation passes the method name and
        parameters to the Java server. The server uses reflection
        to call the requested method. The current implementation
        only supports a limited set of types for parameters and
        return values. Most object types are not supported.

        Allow the driver to close Statement objects when
        $sth->DESTROY is called.

        Fixed, or at least altered, the behavior of the driver
        when the JDBC driver doesn't provide a SQLSTATE value
        in a SQLException. Previously, this problem caused DBI to
        die. 

        Reset the value of the jdbc_error attribute before
        setting a new error.


May 2000
        Version 0.62.

        Changed required version of Convert::BER to 1.31 and
        removed Convert::BER 1.265 from the driver distribution.

        Changed connect implementation to delay creation of $dbh
        until server connection has succeeded. This allows
        checking of $DBI::{err,errstr,state} in a separate
        statement after the call to connect.


January 2000
        Added ability to create a single Server with a
        pre-existing Java connection within a Java
        application. See the example in the example directory. 


        Added $! to socket error reporting; why'd I leave that
        out? 


October 1999
        Initial version.




t/lib.pl
# Default username, password, driver name values for known databases.
sub get_defaults {
    my %defaults;
    $defaults{port} = $ENV{DBDJDBC_PORT} || 43321;
    return undef unless ($ENV{DBDJDBC_URL} or $ENV{DBDJDBC_DRIVER});
    if ($ENV{DBDJDBC_URL}) {
        if ($ENV{DBDJDBC_URL} =~ /^jdbc:opentext:basis/) {
            $defaults{driver}   = "com.opentext.basis.jdbc.BasisDriver";
            $defaults{user}     = $ENV{DBDJDBC_USER} || "user1";
            $defaults{password} = $ENV{DBDJDBC_PASSWORD} || "demo";
        }
        elsif ($ENV{DBDJDBC_URL} =~ /^jdbc:oracle/) {
            $defaults{driver}   = "oracle.jdbc.driver.OracleDriver";
            $defaults{user}     = $ENV{DBDJDBC_USER} || "scott";
            $defaults{password} = $ENV{DBDJDBC_PASSWORD} || "tiger";
        }
        elsif ($ENV{DBDJDBC_URL} =~ /^jdbc:hsqldb/) {
            $defaults{driver}   = "org.hsqldb.jdbcDriver";
            $defaults{user}     = $ENV{DBDJDBC_USER} || "sa";
            $defaults{password} = $ENV{DBDJDBC_PASSWORD} || "";
        }
    }
    elsif ($ENV{DBDJDBC_DRIVER}) {
        $defaults{driver}   = $ENV{DBDJDBC_DRIVER};
        $defaults{user}     = $ENV{DBDJDBC_USER} || "";
        $defaults{password} = $ENV{DBDJDBC_PASSWORD} || "";
    }
    \%defaults;
}



# Taken from Net::Daemon::Test
# Forks a server process.
sub start_server {
    my ($driver, $port) = @_;
    my @cmd = ($ENV{DBDJDBC_JAVA_BIN} || "java", 
               "-Djdbc.drivers=$driver", 
               "-Ddbd.port=$port", 
               "com.vizdom.dbd.jdbc.Server",
               );

    if ($^O =~ /mswin32/i) {
        require Win32;
        require Win32::Process;
        my $proc = $cmd[0];

        # Win32::Process seems to require an absolute path
        my $path;
        my @pdirs;
        if ($proc !~ /\./) {
            $proc .= ".exe";
        }
        if ($proc !~ /^\w\:\\/  &&  $proc !~ /^\\/) {
            # Doesn't look like an absolute path
            foreach my $dir (@pdirs = split(/;/, $ENV{'PATH'})) {
                if (-x "$dir/$proc") {
                    $path = "$dir/$proc";
                    last;
                }
            }
            if (!$path) {
                print STDERR ("Cannot find $proc in the following"
                              , " directories:\n");
                foreach my $dir (@pdirs) {
                    print STDERR "    $dir\n";
                }
                print STDERR "Terminating.\n";
                return undef;
            }
        } else {
            $path = $proc;
        }

        #print "Starting process: proc = $path, args = ", join(" ", @cmd), "\n";
        if (!&Win32::Process::Create($pid, $path,
                                     join(" ", @cmd), 0,
                                     Win32::Process::NORMAL_PRIORITY_CLASS(),
                                     ".")) {
            warn "Cannot create child process: "
                . Win32::FormatMessage(Win32::GetLastError());
            return undef;
        }
        return \$pid;
    }
    else {
        $pid = eval { fork() };
        if (defined($pid)) {
            if (!$pid) {
                # This is the child process, spawn the server.
                exec(@cmd);
            }
            return $pid;
        }
        else {
            return undef;
        }
    }
}

# Kills a server process.
sub stop_server {
    my $server = shift;
    if ($^O =~ /mswin32/i) {
        my $pid = $$server;
        $pid->Kill(0);  # What does this return?
    }
    else {
        kill 1, $server;
    }
}


1;


lib/Bundle/DBD/JDBC.pm
package Bundle::DBD::JDBC;

$VERSION = '0.70';

1;

__END__

=head1 NAME

Bundle::DBD::JDBC - A bundle to install all DBD::JDBC related modules

=head1 SYNOPSIS

C<perl -MCPAN -e 'install Bundle::DBD::JDBC'>

=head1 CONTENTS

Bundle::DBI  - Bundle for DBI by TIMB (Tim Bunce)

Convert::BER - Convert::BER by GBARR (Graham Barr)

DBD::JDBC  - DBD::JDBC by VIZDOM (Gennis Emerson)

=head1 DESCRIPTION

This bundle includes all the modules used by the Perl Database
Interface (DBI) driver for Driver (DBD::JDBC), assuming the
use of DBI version 1.47 or later, created by Tim Bunce.

If you've not previously used the CPAN module to install any
bundles, you will be interrogated during its setup phase.
But when you've done it once, it remembers what you told it.
You could start by running:

  C<perl -MCPAN -e 'install Bundle::CPAN'>

=head1 SEE ALSO

Bundle::DBI

=head1 AUTHOR

Gennis Emerson E<lt>F<gemerson@vizdom.com>E<gt>

=head1 THANKS

This bundle is based on the template included in the DBI::DBD
documentation by Jonathan Leffler
E<lt>F<jleffler@informix.com>E<gt>.

=cut


example/PerlServlet.java
import javax.servlet.*;
import javax.servlet.http.*;

import java.io.*;
import java.util.Hashtable;
import com.vizdom.dbd.jdbc.CgiProcess;

/**
 * This servlet manages a user session and associated JDBC connection,
 * but invokes a Perl script to do the actual database query and 
 * HTML generation.
 */
public class PerlServlet extends HttpServlet 
{
    private String jdbcDriverClass = "com.opentext.basis.jdbc.BasisDriver";
    private String jdbcUrl = "jdbc:opentext:basis://poe:7000/tour_all";
    private String user = "user1";
    private String password = "demo";

    private String perlCommand = "perl d:/testPerlServlet.pl";

    public void init(ServletConfig aConfig) throws ServletException
    {
        super.init(aConfig);
        try
        {
            Class.forName(jdbcDriverClass);
        }
        catch (Exception e)
        {
            System.out.println("Initialization error: " + e.toString());
            throw new ServletException(e);
        }
    }

    public void doPost(HttpServletRequest aRequest, 
        HttpServletResponse aResponse)
        throws ServletException, java.io.IOException
    {
        doGet(aRequest, aResponse);
    }

    public void doGet(HttpServletRequest aRequest, 
        HttpServletResponse aResponse)
        throws ServletException, java.io.IOException
    {
        try
        {
            // Create and cache a JDBC connection. Each user will
            // get their own connection back on each request.
            HttpSession session = aRequest.getSession();
            java.sql.Connection conn = 
                (java.sql.Connection) session.getValue("jdbcConnection");
            if (conn == null)
            {
                conn = java.sql.DriverManager.getConnection(jdbcUrl, user, 
                                                            password);
                session.putValue("jdbcConnection", conn);
            }

            // Use the JDBC connection here, if desired.

            // Create some arguments for the Perl script. These
            // values could obviously come from the HttpRequest.
            Hashtable args = new Hashtable();
            args.put("fields", "eno, name");

            // Exec a Perl script, allowing it to use the JDBC
            // connection.
            CgiProcess cgi = new CgiProcess(perlCommand, args, conn);

            // Ignore any buffering issues and just read each stream.
            String results = getOutput(cgi.getInputStream());
            String errors = getOutput(cgi.getErrorStream());
            int returnValue = cgi.waitFor();
            if (results.equals("") && errors.equals(""))
            {
                String message = (returnValue == 0)
                    ? "No response from " :  "Non-zero return value from ";
                aResponse.sendError(500, message + perlCommand);
                return;
            }

            PrintWriter out = aResponse.getWriter();
            if (!errors.equals(""))
                out.println(errors);
            else
                out.println(results);
        }
        catch (Exception e)
        {
            throw new ServletException(e);
        }
    }

    // This assumes that the output is character, not binary, data.
    private String getOutput(InputStream aStream) throws IOException
    {
        Reader reader = new InputStreamReader(aStream);
        char[] chars = new char[1024];
        java.io.CharArrayWriter buffer = 
            new java.io.CharArrayWriter(5096);
        int count;
        while ((count = reader.read(chars)) != -1)
            buffer.write(chars, 0, count);
        buffer.close();
        return buffer.toString();
    }

}


example/testPerlServlet.pl
# This script is intended to be invoked by a servlet (or another
# Java application). The servlet is responsible for creating a 
# JDBC connection and then creating a com.vizdom.dbd.jdbc.Server
# object which will accept a socket connection from this script.
# The servlet can then exec this script, passing the port and any
# other variables in CGI fashion. This script will, in turn, connect
# to the port and make database calls using the DBI interface
# and the DBD::JDBC module. The servlet will read this script's
# stdout and stderr for the script results and take whatever action
# it desires.

# Since the servlet created the JDBC connection, this script
# doesn't need a JDBC url or database username and password.
# However, since the driver requires them, supply something.

# Setup: the servlet or other application will need to be able
# to use a JDBC driver and exec this script. This script will
# need access to DBD::JDBC (and its supporting modules(s)).


BEGIN
{
    # I needed to set this in order to get Perl scripts to work 
    # when invoked by a servlet.
    $ENV{'SYSTEMROOT'} = "c:\\winnt";
}

use strict;
use CGI;
use DBI;

my $q = new CGI();
my $port = $q->param('com.vizdom.dbd.jdbc.Server.port');

my ($dsn, $dbh, $sth, $row);
$dsn = "dbi:JDBC:hostname=localhost;port=$port;url=PerlServlet";

$dbh = DBI->connect($dsn, undef, undef,
                    { AutoCommit => 1, PrintError => 0, RaiseError => 1})
    or ( die("Failed to connect: ($DBI::err) $DBI::errstr\n"));

# We don't really want to use $q->header; the servlet is handling that.
# Just create HTML for display.

print $q->start_html(-title=>"Perl Script Output");
my $fields = $q->param('fields');
print "The parameter from the servlet is '$fields'.<br><br>\n";

my $query = qq/select $fields from employee where eno < 5000 order by eno/;
$sth = $dbh->prepare($query);
$sth->execute();

print "<table>\n";
print "<tr><th>";
print join("</th><th>", split(/,/, $fields));
print "</th></tr>\n";
print "<tr><td>\n";
while ($row = $sth->fetch()) {
    print "<tr><td>";
    print join("</td><td>", @$row);
    print "</td></tr>\n";
}
print "</table>\n";
print $q->end_html();

$dbh->disconnect();

exit(0);


Makefile.PL
use DBI 1.03;
use DBI::DBD;
use ExtUtils::MakeMaker;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.

WriteMakefile(
        'NAME'  => 'DBD::JDBC',
        'VERSION_FROM' => 'JDBC.pm',
        'NO_META' => 1,
        'dist' => { COMPRESS => 'gzip', SUFFIX => 'gz', },
        'AUTHOR'    => 'Gennis Emerson <gemerson@vizdom.com>',
        'ABSTRACT'  => 'DBI proxy driver using JDBC',
        'PREREQ_PM' => {
           'DBI'     => '1.48',
           'Convert::BER' => 1.31, }, 
); 


package MY;
sub postamble { return main::dbd_postamble(@_); }
sub libscan {
    my ($self, $path) = @_;
    ($path =~ m/\~$/) ? undef : $path;
}
sub test {
    my @cp = ("dbd_jdbc.jar", 
              "t/hsqldb/hsqldb-1.8.0.2.jar", 
              "t/hsqldb/log4j-1.2.13.jar",
              "\$(CLASSPATH)");
    my $inherited = shift->SUPER::test(@_);
    if ($^O eq 'MSWin32') {
        my $cp = join ';', @cp; 
        $inherited =~ s/\$\(FULLPERLRUN\)/set CLASSPATH=$cp\n\t\$\(FULLPERLRUN\)/g;
    }
    else {
        my $cp = join ':', @cp; 
        $inherited =~ s/\$\(FULLPERLRUN\)/CLASSPATH=$cp \$\(FULLPERLRUN\)/g;
    }
    return $inherited;
}

